onnx/cpp2py_export.cc

```diff
@@ -115,13 +115,203 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+      .def_property_readonly("option", &OpSchema::FormalParameter::GetOption)
+      .def_property_readonly("is_homogeneous", &OpSchema::FormalParameter::GetIsHomogeneous)
+      .def_property_readonly("min_arity", &OpSchema::FormalParameter::GetMinArity)
+      .def_property_readonly("differentiation_category", &OpSchema::FormalParameter::GetDifferentiationCategory)
+      // Legacy camel cased names. We retain them for backward compatibility.
+      // TODO(#5074): Remove these before the 1.16 release.
+      .def_property_readonly(
+          "typeStr",
+          [](const OpSchema::FormalParameter& self) {
+            auto warnings = py::module::import("warnings");
+            warnings.attr("warn")(
+                "OpSchema.FormalParameter.typeStr is deprecated. Use OpSchema.FormalParameter.type_str instead.");
+            return self.GetTypeStr();
+          })
+      .def_property_readonly(
+          "isHomogeneous",
+          [](const OpSchema::FormalParameter& self) {
+            auto warnings = py::module::import("warnings");
+            warnings.attr("warn")(
+                "OpSchema.FormalParameter.isHomogeneous is deprecated. Use OpSchema.FormalParameter.is_homogeneous instead.");
+            return self.GetIsHomogeneous();
+          })
+      .def_property_readonly("differentiationCategory", [](const OpSchema::FormalParameter& self) {
+        auto warnings = py::module::import("warnings");
+        warnings.attr("warn")(
+            "OpSchema.FormalParameter.differentiationCategory is deprecated. Use OpSchema.FormalParameter.differentiation_category instead.");
+        return self.GetDifferentiationCategory();
+      });
+
+  op_schema
+      .def(
+          py::init([](std::string name,
+                      std::string domain,
+                      int since_version,
+                      std::string doc,
+                      std::vector<OpSchema::FormalParameter> inputs,
+                      std::vector<OpSchema::FormalParameter> outputs,
+                      std::vector<std::tuple<std::string, std::vector<std::string>, std::string>> type_constraints,
+                      std::vector<OpSchema::Attribute> attributes) {
+            auto self = OpSchema();
+
+            self.SetName(std::move(name)).SetDomain(std::move(domain)).SinceVersion(since_version).SetDoc(doc);
+            // Add inputs and outputs
+            for (auto i = 0; i < inputs.size(); ++i) {
+              self.Input(i, inputs[i]);
+            }
+            for (auto i = 0; i < outputs.size(); ++i) {
+              self.Output(i, outputs[i]);
+            }
+            // Add type constraints
+            for (auto& type_constraint : type_constraints) {
+              std::string type_str;
+              std::vector<std::string> constraints;
+              std::string description;
+              tie(type_str, constraints, description) = type_constraint;
+              self.TypeConstraint(type_str, constraints, description);
```

Comment:
ah, too bad we can't use C++17 yet. but this is a way to avoid copying here.
```suggestion
              tie(type_str, constraints, description) = std::move(type_constraint);
              self.TypeConstraint(std::move(type_str), std::move(constraints), std::move(description));
```

Reply:
Done